src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts

```diff
@@ -276,35 +281,84 @@ export class ObjectExplorerService implements IObjectExplorerService {
 	private async handleSessionCreated(session: azdata.ObjectExplorerSession): Promise<void> {
 		let connection: ConnectionProfile | undefined = undefined;
 		let errorMessage: string | undefined = undefined;
-		if (this._sessions[session.sessionId!]) {
-			connection = this._sessions[session.sessionId!].connection;
+		const sessionId = session.sessionId;
 
-			try {
-				if (session.success && session.rootNode) {
-					let server = this.toTreeNode(session.rootNode, undefined);
-					server.connection = connection;
-					server.session = session;
-					this._activeObjectExplorerNodes[connection!.id] = server;
+
+		await new Promise<void>((resolve, reject) => {
+			const cleanup = () => {
+				if (connection) {
+					this._connectionsWaitingForSession.delete(connection.id);
 				}
-				else {
-					errorMessage = session && session.errorMessage ? session.errorMessage : errSessionCreateFailed;
-					this.logService.error(errorMessage);
+				createNewSessionListener.dispose();
+				clearTimeout(timeoutHandle);
+			}
+			const onTimeout = () => {
+				if (!this._sessions[sessionId]) {
+					this.logService.error(`Timed out waiting for session ${sessionId} to be created.
+					This has probably happened because OE service did not recieve a response for createNewSession from the provider.`);
+					reject(new Error(
+						nls.localize('objectExplorerMissingSession',
+							'Timed out waiting for session {0} to be created. This has probably happened because OE service did not recieve a response for createNewSession from the provider.', sessionId)));
+				} else {
+					this.logService.error(`Timeout waiting for session ${sessionId} to be created for connection "${connection.title}".
```

Comment:
I'm still confused about this else. You say it timed out waiting for the session to be created - but you were able to get a value back from `this._sessions[sessionId]`? Doesn't that imply that a session was found so it succeeded?

Reply:
This bug happens when I am frequently connecting and disconnecting a connection.

This code gets executed whenever we disconnect a OE connection.
https://github.com/microsoft/azuredatastudio/blob/30c22a475ef1a0c51d9eefd919cf06693acb15f0/src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts#L236-L238
Sometimes, close session takes some time to finish. When that happens, this_sessions[sessionId] still has connection property from the previous session.

While we are waiting for closeSession, if I create a new session, and handleSession is called before we get the response of createNewSession this piece of code is not executed..
https://github.com/microsoft/azuredatastudio/blob/30c22a475ef1a0c51d9eefd919cf06693acb15f0/src/sql/workbench/services/objectExplorer/browser/objectExplorerService.ts#L428-L431

Many providers reuse session ids and therefore the value of 
this._session[sessionId].connection is from the old connection and not the new one.

Reply:
@Charles-Gagnon , we can sync up offline if the explanation is not very clear.

Reply:
So it sounds like we don't clear the map until after getting the close session response? 

If that's the case then we should maybe look at clearing the map BEFORE requesting the session be closed - basically the idea being that the user indicated they want to close the session so we should do so on the client side as much as we can and then do a best effort to get the provider to clean the session up themselves as well (which may fail due to other issues, but that shouldn't block us from cleaning up our stuff internally)

Anyways, if that does is something you want to follow up on then I'd suggest creating an issue for it and sending out a separate PR since that's existing behavior and outside the scope of this PR. 

You may want to still consider at least adding a comment explaining this though - it's doesn't seem obvious to me and documenting that may help someone else investigating this in the future.